### PR TITLE
Added `gitops check --pre` command

### PR DESCRIPTION
**Pull request by josecordaz**
Closes: 
- #1103 

**What changed?**
Added `check --pre` command to validate if possible cluster current flux version is compatible with `gitops` and also will validate if Kubernetes version is supported.

If the current flux version is not compatible, it will show:
```
✔ Kubernetes 1.21.1 >=1.19.0-0
✗ Flux 0.0.0 !=~1.21.x
Current flux version is not compatible
```

If the current flux version is compatible it will show:
```
✔ Kubernetes 1.21.1 >=1.19.0-0
✔ Flux 0.21.2 ~=0.21.x
Current flux version is compatible
```

**Why?**
Some users might have flux already installed in their clusters. This would help them to know if it is compatible with gitops.

**How did you test it?**
- Added unit tests

**Release notes**
Yes

**Documentation Changes**
Yes

### CODE REVIEW

File: pkg/services/check/check.go
Comment (josecordaz) [reply pkg/services/check/check.go:30]: @J-Thompson12 I'm confused, what do you mean?

File: pkg/flux/flux.go
Comment (josecordaz) [reply pkg/flux/flux.go:316]: @J-Thompson12 I don't think we are going to have that problem. Why do you think that?

Comment (J-Thompson12) [reply pkg/flux/flux.go:322]: Just curious. I thought this flux check mentioned if there were newer versions available.

File: pkg/services/check/check.go
Comment (J-Thompson12) [reply pkg/services/check/check.go:35]: The check mark in the error text. "✔ Flux is not installed". It seems odd to have a check mark for a fail.

File: pkg/flux/flux.go
Comment (josecordaz) [reply pkg/flux/flux.go:322]: Actually it is, but I'm not printing it out.

File: pkg/services/check/check.go
Comment (josecordaz) [reply pkg/services/check/check.go:35]: @J-Thompson12  Well, the purpose of this new command is to know if a cluster is ready to install gitops on it. Flux not installed is a good scenario, because we are not expecting flux to be installed (at least not yet, there are tickets already defined for that)

File: pkg/services/check/check.go
Comment (palemtnrider) [inline pkg/services/check/check.go:97]: Instead of comparing these as strings, we should consider using the comparisons https://pkg.go.dev/github.com/Masterminds/semver/v3#readme-basic-comparisons in the semver library.

File: pkg/services/check/check_suite_test.go
Comment (palemtnrider) [inline pkg/services/check/check_suite_test.go:12]: Should this be Check Suite instead of Flux Suite?

File: pkg/services/check/check_test.go
Comment (palemtnrider) [inline pkg/services/check/check_test.go:55]:
```suggestion
		Expect(Pre(context, fakeKubeClient, fakeFluxClient, expectedFluxVersion)).To(
			Equal(`✔ Kubernetes 1.21.1 >=1.19.0-0
✔ Flux is not installed`))
```
NIt- could be simplified to a single Expect() call

File: pkg/services/check/check_test.go
Comment (palemtnrider) [inline pkg/services/check/check_test.go:95]:
```suggestion
	It("should fail when expected flux version is not valid", func() {
```

File: test/acceptance/test/add_tests.go
Comment (palemtnrider) [inline test/acceptance/test/add_tests.go:174]: Is it intentional that this test is focused on this test case?

**Review by palemtnrider — CHANGES_REQUESTED** [review]
Good to see the addition of the check command.  

Are there plans for install to execute these checks?

Just a handful of suggestions prior to merging.

File: cmd/gitops/check/cmd.go
Comment (josecordaz) [reply cmd/gitops/check/cmd.go]: There is no changes in the output when this flag is not present:

Without cobra.CheckErr
```
./bin/gitops check
Error: required flag(s) "pre" not set
```

With cobra.CheckErr
```
./bin/gitops check
Error: required flag(s) "pre" not set
```

🤔

File: cmd/gitops/check/cmd.go
Comment (josecordaz) [reply cmd/gitops/check/cmd.go:36]: The ticket didn't specify that, I wonder if we should do this in another ticket.

File: cmd/gitops/check/cmd.go
Comment (josecordaz) [reply cmd/gitops/check/cmd.go:64]: Sounds good. This change will add a new line to the output. What would you like me to put instead of ERROR_MESSAGE?
```
✔ Kubernetes 1.21.1 >=1.19.0-0
✗ Flux 0.19.0 !=0.21.0
Current flux version is not compatible
Error: ERROR_MESSAGE
```

File: pkg/services/check/check.go
Comment (josecordaz) [reply pkg/services/check/check.go:74]: This is not is not in install path yet, maybe you got confused with PR #1002 . But I agree let's use that function instead.

Comment (josecordaz) [reply pkg/services/check/check.go:74]: I'm having second thoughts.  I was adding the method `FetchNamespaceWithLabel` from PR #1002 . But then I realized I would need another method to get the label values. So I think the current method `GetNamespaces` allows me to do both at the same time. What do you thing about this?

File: pkg/services/check/check.go
Comment (josecordaz) [reply pkg/services/check/check.go:97]: There is no need for that as we are just interesting in if they are equal or not. If we were to compare if one is less or grater then the other one then it would make more sense.

File: pkg/services/check/check_suite_test.go
Comment (josecordaz) [reply pkg/services/check/check_suite_test.go:12]: yes

File: pkg/services/check/check_test.go
Comment (josecordaz) [reply pkg/services/check/check_test.go:55]: My personal preference is to have the function call and the Expect in the same level of indentation. Do this approach also checks that error is not happening? I couldn't find a docs explaining. For this particular case actually it would be cleaner to move the expected output to the first level of indentation as well so it's more readable.

File: test/acceptance/test/add_tests.go
Comment (josecordaz) [reply test/acceptance/test/add_tests.go]: good catch

**Review by palemtnrider — COMMENTED** [review]
Looking good.  Just a few more minor suggestions.

File: cmd/gitops/check/cmd.go
Comment (J-Thompson12) [inline cmd/gitops/check/cmd.go:56]: I am confused, I thought with marks comments that check should run more than just the pre checks. I was under the impression that if I run `gitops check` it checks everything and then if I add --pre it will only run the pre check. I am not seeing anywhere that the pre flag is even being used.

File: pkg/services/check/check.go
Comment (palemtnrider) [reply pkg/services/check/check.go:75]: Sorry, my message wasn't clear.  Can we use the FetchNamespaceWithLabel() https://github.com/weaveworks/weave-gitops/blob/390fc7c3db7454d30be430459c8fec9997ceb24f/pkg/kube/kubehttp.go#L494 instead of duplicating the logic here?

**Review by palemtnrider — CHANGES_REQUESTED** [review]
Just one additional suggestion to use the existing FetchNamespaceWithLabel() other than that LGTM

File: pkg/services/check/check.go
Comment (palemtnrider) [reply pkg/services/check/check.go:75]: Looks like this new function just landed in main so when you rebase you will pick it up.

Comment (josecordaz) [reply pkg/services/check/check.go:75]: @palemtnrider  That method only returns a namespace name. I need to get a specific label's value. There are two labels values I'm interested in: `app.kubernetes.io/part-of` and `app.kubernetes.io/version`. I am using the first label to find the right namespace. Once I have it. I can get the value for the second label. So then I can continue with the logic. Let me know if I'm missing something.

Comment (palemtnrider) [reply pkg/services/check/check.go:75]: Sorry, I should have picked up on that.  🤦

**Review by palemtnrider — APPROVED** [review]
LGTM

**Comment by palemtnrider** [comment]
Well done @josecordaz 👍

File: cmd/gitops/check/cmd.go
Comment (josecordaz) [reply cmd/gitops/check/cmd.go:56]: Good question. This ticket is just about running pre-checks. Any other checks are out of the scope of this ticket.

**Review by J-Thompson12 — APPROVED** [review]
LGTM

File: pkg/services/check/check.go
Comment (jpellizzari) [inline pkg/services/check/check.go:47]: @josecordaz We need a newline character here. This is what I get:

```
jordan@UbuntuDesktop:~/Projects/weave-gitops$ ./bin/gitops check --pre
✔ Kubernetes 1.21.1 >=1.19.0-0
✗ Flux 0.21.0 !=~0.24.x
Current flux version is not compatiblejordan@UbuntuDesktop:~/Projects/gitopstest/weave-gitops$ 
```

What I would like is:

```
jordan@UbuntuDesktop:~/Projects/weave-gitops$ ./bin/gitops check --pre
✔ Kubernetes 1.21.1 >=1.19.0-0
✗ Flux 0.21.0 !=~0.24.x
Current flux version is not compatible
jordan@UbuntuDesktop:~/Projects/gitopstest/weave-gitops$ 
```

**Review by jpellizzari — APPROVED** [review]
Approving with a caveat: one more newline character is needed. Once that is fixed, feel free to merge.

**Pull request merged**
